independent-projects/arc/runtime/src/main/java/io/quarkus/arc/RemovedBean.java

```diff
@@ -21,12 +21,35 @@
 
     /**
      * @return the bean types
+     * @deprecated use {@link #types} to allow for future optimisations.
      */
-    public Set<Type> getTypes();
+    @Deprecated(forRemoval = true)
+    Set<Type> getTypes();
```

Comment:
I tend to disagree with this modification. A Set is much more useful than an Iterable and corresponds to the `javax.enterprise.inject.spi.BeanAttributes.getTypes()` methods.

Maybe we could remove the `qualifiers()` and `types()` and change the `boolean matchesType(Type requiredType)` signature to `boolean matches(Type requiredType, Annotation... qualifiers)`. This method could be optimized and used in `ArcContainer` whereas the `getTypes()` and `getQualifiers()` could possibly create a new set for every invocation from any data structure used internally.

Reply:
Well but think of your clients.. this code is only used by that little utility which dumps a nice error message when a removed bean shouldn't have been removed: we surely can handle a little extra complexity there since it's not "public" ?

Reply:
BTW regarding style.. ArC is consistently exposing Set(s) which often requires wrapping in immutable sets, sometimes it requires full copies. And iterators aren't free either.

In general it would be more efficient to internalize the logic into the host which controls the collection, like I did by introducing the specialized `matchesType` method here.

The drawback is of course the specialization.. far less flexible and it only works if you know very well what the purpose of the iteration is going to be.

So I tend to make such changes for internal/private components, but sure I understand you can't do it for the ArC public APIs as a general rule. Still useful to consider exposing some specialized methods though, for the most common cases can be implemented more efficiently.